### ISSUE TITLE
display-name has false positive on nested use of memo and forwardRef

### ISSUE THREAD

**Issue by fefrei** [jsx-eslint/eslint-plugin-react#2954]
Since #2399, `eslint-plugin-react` knows not to complain about code like this:

```tsx
export const ComponentWithMemo = React.memo(function Component({ world }) {
 return <div>Hello {world}</div>
})

export const ComponentWithForwardRef = React.forwardRef(function Component({ world }) {
  return <div>Hello {world}</div>
})
```

In that pull request, nested use of `memo` and `forwardRef` like so...

```tsx
export const ComponentWithForwardRef = React.memo(React.forwardRef(function Component({ world }) {
  return <div>Hello {world}</div>
}))
```

...was intentionally treated as a `display-name` violation due to [a React bug causing such components to be anonymous](https://github.com/facebook/react/issues/16722). This issue was [solved in November 2019](https://github.com/facebook/react/pull/17274), so the `display-name` should now accept this.

**Comment by ljharb** [comment]
In which React versions is that fixed in React? We should only consider that satisfactory in those versions.

**Comment by fefrei** [comment]
I'm unsure, but judging by the date, it should have gone in somewhere in the React 16 cycle.

**Comment by ljharb** [comment]
Clicking on [the commit that merged the PR](https://github.com/facebook/react/commit/4f02c93c7cb213355b3c7d2011e531431d1c6dd8) implies it's actually gone out in v0.14.10, v15.7.0, and v16.12.0+ - so that's the range we should be using.

**Comment by ljharb** [comment]
`React.memo` takes a component, so a function passed to it absolutely should be required to have a displayName, since it won't implicitly infer a name from anywhere.

`React.forwardRef` does not take a component, so this shouldn't complain.

However, the specific nested case should complain - unless React is one of the versions mentioned [here](https://github.com/jsx-eslint/eslint-plugin-react/issues/2954#issuecomment-806199006), in which case it should not complain.

To handle that in the code, the `report` call should only be made if the react version fetched from settings is high enough.
